File: data/master_dataset/extracted/DARTIS_2019_dataset/docs/examples/lookup.py
"""DARTIS_2019 - filename -> metadata lookup (Python / FastAPI reference).

Unzip the package, set DATASET_ROOT, and call lookup_by_filename() with the
name of the file a user uploaded.
"""
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

DATASET_ROOT = Path(os.environ.get("DATASET_ROOT", "./DARTIS_2019_dataset"))

# Load once at boot: ~1 MB, gives O(1) lookup for every accepted name.
with open(DATASET_ROOT / "metadata" / "filename_lookup.json", encoding="utf-8") as fh:
    ALIASES: dict[str, str] = json.load(fh)

IMAGE_EXTENSIONS = (".jpg", ".jpeg", ".png", ".tif", ".tiff")


def normalise(name: str) -> str:
    """Strip directories, lowercase, trim."""
    return os.path.basename(str(name).replace("\\", "/")).strip().lower()


def resolve_id(filename: str) -> str | None:
    """Resolve an uploaded filename to a record id, tolerating a different extension."""
    n = normalise(filename)
    if n in ALIASES:
        return ALIASES[n]
    stem = os.path.splitext(n)[0]
    if stem in ALIASES:
        return ALIASES[stem]
    for ext in IMAGE_EXTENSIONS:
        if stem + ext in ALIASES:
            return ALIASES[stem + ext]
    return None


def lookup_by_filename(filename: str) -> dict[str, Any] | None:
    """Return the full record for an uploaded filename, or None if unmatched."""
    rid = resolve_id(filename)
    if rid is None:
        return None
    with open(DATASET_ROOT / "metadata" / "records" / f"{rid}.json", encoding="utf-8") as fh:
        record = json.load(fh)
    record["absolute_image_path"] = str(DATASET_ROOT / record["files"]["image"])
    record["absolute_annotation_path"] = (
        str(DATASET_ROOT / record["files"]["annotation"]) if record["files"]["annotation"] else None
    )
    return record


def lookup_by_sha256(digest: str) -> dict[str, Any] | None:
    """Content-based fallback: match an upload by SHA-256 even if it was renamed.

    Build the index once (it is not shipped, to keep the package small):
        BY_HASH = {r["files"]["image_sha256"]: r["record_id"] for r in iter_records()}
    """
    for path in (DATASET_ROOT / "metadata" / "records").glob("*.json"):
        with open(path, encoding="utf-8") as fh:
            record = json.load(fh)
        if record["files"].get("image_sha256") == digest.lower():
            return record
    return None


# ---- FastAPI endpoint -------------------------------------------------------
# from fastapi import FastAPI, File, UploadFile, HTTPException
#
# app = FastAPI()
#
# @app.post("/api/lookup")
# async def lookup(image: UploadFile = File(...)):
#     record = lookup_by_filename(image.filename)
#     if record is None:
#         raise HTTPException(status_code=404, detail=f"No dataset match for {image.filename}")
#     return {"found": True, "query": image.filename, "record": record}


if __name__ == "__main__":
    import sys

    for arg in sys.argv[1:]:
        rec = lookup_by_filename(arg)
        print(json.dumps(rec, indent=2) if rec else f"no match: {arg}")

File: data/master_dataset/extracted/DARTIS_2019_dataset/docs/examples/lookup.js
/**
 * DARTIS_2019 — filename -> metadata lookup (Node.js / Express reference).
 *
 * Unzip the package, point DATASET_ROOT at it, and call lookupByFilename()
 * with the name of the file a user uploaded.
 */
const fs = require("fs");
const path = require("path");

const DATASET_ROOT = process.env.DATASET_ROOT || "./DARTIS_2019_dataset";

// Load once at boot: ~1 MB, gives O(1) lookup for every accepted name.
const aliases = JSON.parse(
  fs.readFileSync(path.join(DATASET_ROOT, "metadata", "filename_lookup.json"), "utf8")
);

/** Normalise any uploaded name: strip directories, lowercase, trim. */
function normalise(name) {
  return String(name).split(/[\\/]/).pop().trim().toLowerCase();
}

/** Resolve an uploaded filename to a record id, tolerating a missing/other extension. */
function resolveId(filename) {
  const n = normalise(filename);
  if (aliases[n]) return aliases[n];
  const stem = n.replace(/\.[^.]+$/, "");
  if (aliases[stem]) return aliases[stem];
  for (const ext of [".jpg", ".jpeg", ".png", ".tif", ".tiff"]) {
    if (aliases[stem + ext]) return aliases[stem + ext];
  }
  return null;
}

/** Full record for an uploaded filename, or null when the dataset has no match. */
function lookupByFilename(filename) {
  const id = resolveId(filename);
  if (!id) return null;
  const record = JSON.parse(
    fs.readFileSync(path.join(DATASET_ROOT, "metadata", "records", `${id}.json`), "utf8")
  );
  return {
    ...record,
    absolute_image_path: path.join(DATASET_ROOT, record.files.image),
    absolute_annotation_path: record.files.annotation
      ? path.join(DATASET_ROOT, record.files.annotation)
      : null,
  };
}

// ---- Express endpoint -------------------------------------------------------
// const express = require("express");
// const multer = require("multer");
// const app = express();
// const upload = multer({ storage: multer.memoryStorage() });
//
// app.post("/api/lookup", upload.single("image"), (req, res) => {
//   const name = req.file ? req.file.originalname : req.body.filename;
//   const record = lookupByFilename(name);
//   if (!record) return res.status(404).json({ found: false, query: name });
//   res.json({ found: true, query: name, record });
// });
//
// app.listen(3000);

module.exports = { lookupByFilename, resolveId, normalise };
